### PR TITLE
Adds 1-RDM feature to MPS.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds RDM feature to MPS so that we can implement bitstring sampling properly in #460 .

At a high level the algorithm works, by contracting all blocks to the right of the rdm index and then contracting all blocks to the left of the rdm and index and then slowly contracting them into the index block. This keeps intermediate tensor size down.

cc: @jaeyoo @dstrain115 @rmlarose

### CODE REVIEW

File: lib/mps_statespace.h [hunk 1]
```diff
@@ -232,6 +233,149 @@ class MPSStateSpace {
...
+    // The [bond_dim, bond_dim] block in state_raw now contains the contraction
+    // up to, but not including index.
+    // Contract rightmost blocks.
+    offset = GetBlockOffset(state, num_qubits - 1);
+    new (&top) ConstMatrixMap((Complex*)(scratch_raw + offset), bond_dim, 2);
+    new (&bot) ConstMatrixMap((Complex*)(state_raw + offset), bond_dim, 2);
+    new (&partial_contract)
+        MatrixMap((Complex*)(scratch_raw + end), bond_dim, bond_dim);
+    new (&partial_contract2)
+        MatrixMap((Complex*)(scratch_raw + end + 2 * bond_dim * bond_dim),
```
Comment (sergeisakov) [inline lib/mps_statespace.h:299]: Set a variable to `(Complex*)(scratch_raw + end + 2 * bond_dim * bond_dim)`? This expression is used four times in this function.

Comment (sergeisakov) [inline lib/mps_statespace.h:299]: Nit: set a variable to `(Complex*)(scratch_raw + end + 2 * bond_dim * bond_dim)`? This expression is used four times in this function.

File: lib/mps_statespace.h [hunk 2]
```diff
@@ -232,6 +233,149 @@ class MPSStateSpace {
...
+                  bond_dim, 2 * bond_dim);
+
+    partial_contract.setZero();
+    partial_contract(0, 0) = 1;
+    if (index < num_qubits - 1) {
+      partial_contract.noalias() = top * bot.adjoint();
+    }
+
+    for (unsigned i = num_qubits - 2; i > index; i--) {
```
Comment (sergeisakov) [inline lib/mps_statespace.h:308]: Nit: use `--i` for consistency.

**Review by sergeisakov — APPROVED** [review]
LGTM

File: tests/mps_statespace_test.cc [hunk 3]
```diff
@@ -586,6 +586,321 @@ TEST(MPSStateSpaceTest, InnerProduct4) {
...
+  //        |           |     |     |
+  //        |           |     |     |
+  //      +-+-+ +-+-+ +-+-+ +-+-+ +-+-+
+  // mps  | 0 +-+ 1 +-+ 2 +-+ 3 |-+ 4 |
+  //      +---+ +---+ +---+ +---+ +---+
+  //              |
+  //-----------------------------------------
+  // And so on.
+
+  float *rdm = new float[8];
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tests/mps_statespace_test.cc]: Changed.

File: lib/mps_statespace.h [hunk 4]
```diff
@@ -232,6 +233,149 @@ class MPSStateSpace {
     return partial_contract(0, 0);
   }
 
+  // Compute the 2x2 1-RDM of state on index. Result written to rdm.
+  // Requires: scratch to be allocated.
+  static void ReduceDensityMatrix(MPS& state, MPS& scratch, int index,
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply lib/mps_statespace.h:238]: Yes

File: lib/mps_statespace.h [hunk 5]
```diff
@@ -232,6 +233,149 @@ class MPSStateSpace {
...
+    fp_type* scratch_raw = scratch.get();
+
+    Copy(state, scratch);
+
+    // Contract leftmost blocks together, store result in state scratch.
+    ConstMatrixMap top((Complex*)scratch_raw, 2, bond_dim);
+    ConstMatrixMap bot((Complex*)state_raw, 2, bond_dim);
+    MatrixMap partial_contract((Complex*)(state_raw + end), bond_dim, bond_dim);
+    MatrixMap partial_contract2(
+        (Complex*)(state_raw + end + 2 * bond_dim * bond_dim), bond_dim,
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 1 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 2 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

**Pull request merged**